flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -0,0 +1,422 @@
+	 * @param expectedVersion Expected version of the node to replace
+	 * @param state           The new state to replace the old one
+	 */
+	public void replace(String pathInZooKeeper, int expectedVersion, T state) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+		checkNotNull(state, "State");
+
+		StateHandle<T> oldStateHandle = get(pathInZooKeeper);
+
+		StateHandle<T> stateHandle = stateHandleProvider.createStateHandle(state);
+
+		boolean success = false;
+
+		try {
+			// Serialize the new state handle. This writes the state to the backend.
+			byte[] serializedStateHandle = InstantiationUtil.serializeObject(stateHandle);
+
+			// Replace state handle in ZooKeeper.
+			client.setData()
+					.withVersion(expectedVersion)
+					.forPath(pathInZooKeeper, serializedStateHandle);
+
+			success = true;
+		}
+		finally {
+			if (success) {
+				oldStateHandle.discardState();
+			}
+			else {
+				stateHandle.discardState();
+			}
+		}
+	}
+
+	/**
+	 * Returns the version of the node if it exists or <code>-1</code> if it doesn't.
+	 *
+	 * @param pathInZooKeeper Path in ZooKeeper to check
+	 * @return Version of the ZNode if the path exists, <code>-1</code> otherwise.
+	 * @throws Exception
+	 */
+	public int exists(String pathInZooKeeper) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+
+		Stat stat = client.checkExists().forPath(pathInZooKeeper);
+
+		if (stat != null) {
+			return stat.getVersion();
+		}
+
+		return -1;
+	}
+
+	/**
+	 * Gets a state handle from ZooKeeper.
+	 *
+	 * @param pathInZooKeeper Path in ZooKeeper to get the state handle from (expected to
+	 *                        exist and start with a '/').
+	 * @return The state handle
+	 * @throws Exception
+	 */
+	@SuppressWarnings("unchecked")
+	public StateHandle<T> get(String pathInZooKeeper) throws Exception {
+		checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
+
+		byte[] data = client.getData().forPath(pathInZooKeeper);
+
+		return (StateHandle<T>) InstantiationUtil.deserializeObject(data, ClassLoader
+				.getSystemClassLoader());
+	}
+
+	/**
+	 * Gets all available state handles from ZooKeeper.
+	 *
+	 * @return All state handles from ZooKeeper.
+	 * @throws ConcurrentModificationException
+	 */
+	@SuppressWarnings("unchecked")
+	public List<ZooKeeperStateHandle<T>> getAll() throws Exception {
+		// Initial cVersion (number of changes to the children of this node)
+		int initialCVersion = client.checkExists().forPath("/").getCversion();
+
+		List<String> children = client.getChildren().forPath("/");
+
+		final List<ZooKeeperStateHandle<T>> stateHandles = new ArrayList<>(children.size());
+
+		for (String path : children) {
+			path = "/" + path;
+
+			try {
+				final StateHandle<T> stateHandle = get(path);
+				stateHandles.add(new ZooKeeperStateHandle(stateHandle, path));
+			}
+			catch (KeeperException.NoNodeException e) {
+				throw new ConcurrentModificationException("Modification during getAll()", e);
```

Comment:
This indicates a corner case where another job manager is modifying the stat or a removal overtaking this operation. I think it's better to fail clearly here.